speakerbox/main.py

```diff
@@ -0,0 +1,202 @@
+        and cast the "label" column into ClassLabels and the "audio" column into Audio.
+        If provided a Dataset object (or the converted pandas DataFrame),
+        this will convert it into a DatasetDict with
+        random train, test, and validation splits.
+    model_name: str
+        A name for the model. This will also create a directory with the same name
+        to store the produced model in.
+        Default: "trained-speakerbox"
+    model_base: str
+        The model base to use before fine tuning.
+    max_duration: float
+        The maximum duration to use for each audio clip.
+        Any clips longer than this will be trimmed.
+        Default: 2.0
+    batch_size: int
+        The number of examples to use in a batch during training.
+        Default: 2
+
+    Returns
+    -------
+    model_storage_path: Path
+        The path to the directory where the model is stored.
+    """
+    # Load feature extractor
+    feature_extractor = Wav2Vec2FeatureExtractor.from_pretrained(model_base)
+
+    # Convert provided dataset to HF Dataset
+    if isinstance(dataset, pd.DataFrame):
+        dataset = Dataset.from_pandas(dataset, preserve_index=False)
+        dataset = dataset.class_encode_column("label")
+        dataset = dataset.cast_column("audio", Audio(feature_extractor.sampling_rate))
+
+    # Convert into train, test, and validate dict
+    if isinstance(dataset, Dataset):
+        train_and_test = dataset.train_test_split(test_size=0.4)
+        test_and_valid = train_and_test["test"].train_test_split(test_size=0.5)
+        ds_dict = DatasetDict(
+            {
+                "train": train_and_test["train"],
+                "test": test_and_valid["train"],
+                "valid": test_and_valid["test"],
+            }
+        )
+
+    # Show dataset summary stats
+    for subset in ["train", "test", "valid"]:
+        pd_subset = ds_dict[subset].to_pandas()
+        print(f"Summary stats for '{subset}' dataset")
+        print(f"n-labels: {pd_subset.label.nunique()}")
+        print(f"Avg duration: {pd_subset.duration.mean()}")
+        print(f"Min duration: {pd_subset.duration.min()}")
+        print(f"Max duration: {pd_subset.duration.max()}")
+        print(f"StD duration: {pd_subset.duration.std()}")
+        print("-" * 80)
+
+    # Construct label to id and vice-versa LUTs
+    label2id, id2label = dict(), dict()
+    for i, label in enumerate(ds_dict["train"].features["label"].names):
+        label2id[label] = str(i)
+        id2label[str(i)] = label
+
+    # Construct preprocessing function
+    def preprocess(
+        examples: arrow_dataset.Batch,
+    ) -> feature_extraction_utils.BatchFeature:
+        audio_arrays = [x["array"] for x in examples["audio"]]
+        inputs = feature_extractor(
+            audio_arrays,
+            sampling_rate=feature_extractor.sampling_rate,
+            max_length=int(feature_extractor.sampling_rate * max_duration),
+            do_normalize=True,
```

Comment:
How are the mean and variance calculated and used in this normalization step? Are the mean and variance calculated per batch and used to normalize that batch? Usually, the mean and variance are calculated from the train set and used to normalize the entire set.

Or is the mean and variance calculated from the entire set and then used to normalize the entire set? If yes then there is data leakage from holdout validation set to training step -- the training is peeking at the mean and variance of the validation set to pick the best model.

Reply:
Great question and to be honest, I don't entirely know. The feature extractor comes shipped with the model so we would have to find their original code. This could just be normalizing a single audio array though. The mean of the audio array freq and putting between 0 and 1 or something? It doesn't _have_ to take into account the whole dataset. But I see your point.

Reply:
Yea normalization (according to the doc) is transforming the input's values to be between 0 and 1, by subtracting the mean and dividing the variance from the input. It's important where the mean and variance are calculated from. I don't know what the effects on training would be if the mean and variance are calculated for each input and used only for that input. I've only seen the mean and variance calculated from the train set, those values are frozen and used to normalize (make the values between 0 and 1) the entire set.

Reply:
Thinking about this now, I think the current code is fine. I think maybe the mean and variance are calculated from the original (non-cdp) dataset.

Reply:
I can try training a model with it turned off and see what effect if has on the accuracy

Reply:
Looks like it's calculating mean and variance from each input and not using the mean and variance of the original dataset

https://github.com/huggingface/transformers/blob/v4.16.2/src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py#L99

Reply:
Could you try turning off normalization, but perform normalization ourselves?

Calculate the mean and variance from the train set, normalize the entire set?

Reply:
At first glance the accuracy does seem high, but IMHO we are classifying only a few different speakers from several hours of data. I don't think this number looks too suspicious considering speaking voices are very distinct. It looks like you have done your best to balance out speakers in each set and even if accuracy is not the best metric here if we wanted to use any of the F score family of metrics we would have to assign negatives and positives.

Reply:
Worth noting that all of our miss classifications are only female. Seattle does have 2/3 female so by default if we miss classify there is a higher probability it would be female all things being equal. From 6 misses on your confusion matrix a few females had 2 miss classifications. I am simply bringing it up so if we do say King County next where females have 1/3 if we only miss female members again we may want to look deeper.

Reply:
Wondering if the miss-classes had less time in meetings? Gonzalez is new so I am guessing that could be true at least for one?

Reply:
I balanced the whole dataset to have n number of samples where n was the number with the least number of samples per person. So they all had a random ~200 samples or so iirc.

Gonzalez in the un-balanced dataset is actually the highest amount of data as I think you simply confused her with someone else. Gonzalez used to be the council president. But yes, will be interesting to test in king county